data-prepper-event/src/main/java/org/opensearch/dataprepper/core/event/CachingEventKeyFactory.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.dataprepper.core.event;
+
+import com.github.benmanes.caffeine.cache.Cache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+import org.opensearch.dataprepper.model.event.EventKey;
+import org.opensearch.dataprepper.model.event.EventKeyFactory;
+import org.springframework.context.annotation.Primary;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+import java.util.Arrays;
+import java.util.Objects;
+
+@Named
+@Primary
+class CachingEventKeyFactory implements EventKeyFactory {
+    private final EventKeyFactory delegateEventKeyFactory;
+    private final Cache<CacheKey, EventKey> cache;
+
+    private static class CacheKey {
+        private final String key;
+        private final EventAction[] eventActions;
+
+        private CacheKey(final String key, final EventAction[] eventActions) {
+            this.key = key;
+            this.eventActions = eventActions;
+        }
+
+        @Override
+        public boolean equals(final Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+            final CacheKey cacheKey = (CacheKey) o;
+            return Objects.equals(key, cacheKey.key) && Arrays.equals(eventActions, cacheKey.eventActions);
+        }
+
+        @Override
+        public int hashCode() {
+            int result = Objects.hash(key);
+            result = 31 * result + Arrays.hashCode(eventActions);
+            return result;
+        }
+    }
+
+    @Inject
+    CachingEventKeyFactory(final EventKeyFactory delegateEventKeyFactory, final EventConfiguration eventConfiguration) {
+        this.delegateEventKeyFactory = delegateEventKeyFactory;
+        cache = Caffeine.newBuilder()
```

Comment:
Nice! I was not familiar with this library before

Reply:
Caffeine is preferred over the Guava cache. And we actually have a Checkstyle to prevent use of the Guava cache.